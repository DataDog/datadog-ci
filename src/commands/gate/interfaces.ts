import {Writable} from 'stream'

import type {AxiosPromise} from 'axios'

import {SpanTags} from '../../helpers/interfaces'

export interface Payload {
  requestId: string
  startTimeMs: number
  spanTags: SpanTags
  userScope: Record<string, string[]>
  options: PayloadOptions
}

export interface PayloadOptions {
  dryRun: boolean
  noWait: boolean
  isLastRetry?: boolean
}

export interface EvaluationResponsePayload {
  data: {
    attributes: EvaluationResponse
  }
}

export interface EvaluationResponse {
  status: string
  rule_evaluations: RuleEvaluation[]
  metadata?: {
    wait_time_ms: number
  }
}

export interface RuleEvaluation {
  rule_id: string
  rule_name: string
  status: string
  is_blocking: boolean
  failure_reason: string
  details_url: string
}

export interface APIHelper {
  evaluateGateRules(evaluateRequest: Payload, write: Writable['write']): AxiosPromise<EvaluationResponsePayload>
}
