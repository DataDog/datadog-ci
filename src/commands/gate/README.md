# GATE command

Evaluate Quality Gates rules in Datadog. The exit status of the command will change depending on the result of the gate evaluation.

## Usage

### Commands

#### `evaluate`

The `evaluate` command evaluates the matching Quality Gates rules in Datadog.

```bash
datadog-ci gate evaluate [--scope] [--tags] [--dry-run] [--fail-if-unavailable] [--fail-on-empty] [--no-wait]
```

For example:

```bash
datadog-ci gate evaluate --scope team:backend --scope team:frontend --fail-on-empty
```

- `--scope` is an array of key value pairs of the form `key:value`. This parameter sets additional scope when retrieving matching rules. Only the rules matching the scope provided will be evaluated.
- `--tags` is an array of key value pairs of the form `key:value`. This parameter sets global tags that are applied to all results. The upload process merges the tags passed on the command line with the tags in the `DD_TAGS` environment variable. If a key appears in both `--tags` and `DD_TAGS`, the value in `DD_TAGS` takes precedence.
- `--dry-run` runs the command without the final evaluation step. All other checks are performed. The default value is `false`.
- `--fail-if-unavailable` fails the command if Datadog is unavailable. The default value is `false`.
- `--fail-on-empty` fails the command if no matching rules are found in Datadog. The default value is `false`.
- `--timeout` is the command timeout in seconds. The default value is `600`.
- `--no-wait` removes the waiting time (30s) that is in place to ensure that events (for example: tests) can be properly queried by the rules. This mechanism prevents rules from being incorrectly evaluated due to the absence of events. Read more on the quality gates documentation before passing this flag. The default value is `false`.


### Environment variables

Additionally, you can configure the `gate` command with the following environment variables:

- `DD_API_KEY` (**required**): The API key used to authenticate the requests.
- `DD_APP_KEY` (**required**): The application key used to authenticate the requests.
- `DD_TAGS`: Sets global tags applied to all spans. The format must be `key1:value1,key2:value2`. The upload process merges the tags passed on the command line with the tags in the `--tags` parameter. If a key appears in both `--tags` and `DD_TAGS`, the value in `DD_TAGS` takes precedence.
- `DD_SITE`: Your Datadog site, for example, datadoghq.com or datadoghq.eu.

### Dependencies

- [`git`](https://git-scm.com/downloads) is used for extracting repository metadata.

### End-to-end testing process

To verify the command works as expected, use `--dry-run`:

```bash
export DD_API_KEY='<API key>'
export DD_APP_KEY='<App key>'

yarn launch gate evaluate --scope team:backend --dry-run
```

Successful output looks like the example below:

```bash
ℹ️ Evaluating rules matching the following information:
Repository: git@github.com:DataDog/datadog-ci.git
Branch: master
team: backend

Dry run mode is enabled. Not evaluating the rules.
```

## Further reading

Additional helpful documentation, links, and articles:

- [Learn about Quality Gates][1]

[1]: https://docs.datadoghq.com/quality_gates/
