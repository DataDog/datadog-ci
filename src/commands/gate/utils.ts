import {getCommonAppBaseURL} from '../../helpers/app'

export const getBaseUrl = () => {
  const site = process.env.DD_SITE || 'datadoghq.com'
  const subdomain = process.env.DD_SUBDOMAIN || ''

  return getCommonAppBaseURL(site, subdomain)
}

export const getBaseIntakeUrl = () => {
  if (process.env.DD_SITE) {
    return `https://quality-gates.${process.env.DD_SITE}`
  }

  return 'https://quality-gates.datadoghq.com'
}

/**
 * Receives an array of the form ['key:value', 'key2:value2_1', 'key2:value2_2']
 * and returns an object of the form {key: ['value'], key2: ['value2_1, value2_2']}
 */
export const parseScope = (scope: string[]) => {
  try {
    return scope.reduce((acc: {[key: string]: string[]}, keyValuePair) => {
      if (!keyValuePair.includes(':')) {
        return acc
      }

      const firstColon = keyValuePair.indexOf(':')
      const key = keyValuePair.substring(0, firstColon)
      const value = keyValuePair.substring(firstColon + 1)

      if (acc.hasOwnProperty(key)) {
        if (!acc[key].includes(value)) {
          acc[key].push(value)
        }
      } else {
        acc[key] = [value]
      }

      return acc
    }, {})
  } catch (e) {
    return {}
  }
}

export const getStatus = (error: any) => {
  return error.response?.status
}

export const is4xxError = (error: any) => {
  const status = getStatus(error)

  return status && status >= 400 && status <= 499
}

export const isBadRequestError = (error: any) => {
  const status = getStatus(error)

  return status && status === 400
}

export const is5xxError = (error: any) => {
  const status = getStatus(error)

  return status && status >= 500 && status <= 599
}

export const isTimeout = (error: any) => {
  return error.message === 'wait'
}
