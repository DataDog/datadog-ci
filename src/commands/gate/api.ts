import type {EvaluationResponsePayload, Payload} from './interfaces'
import type {AxiosPromise, AxiosRequestConfig} from 'axios'
import type {Writable} from 'stream'

import {getRequestBuilder} from '../../helpers/utils'

export const evaluateGateRules = (
  request: (args: AxiosRequestConfig) => AxiosPromise<EvaluationResponsePayload>
) => async (evaluateRequest: Payload, write: Writable['write']) => {
  const payload = JSON.stringify({
    data: {
      id: evaluateRequest.requestId,
      type: 'gate_evaluation',
      attributes: {
        tags: evaluateRequest.spanTags,
        user_scope: evaluateRequest.userScope,
        start_time_ms: evaluateRequest.startTimeMs,
        options: {
          no_wait: evaluateRequest.options.noWait,
          dry_run: evaluateRequest.options.dryRun,
        },
      },
    },
  })

  return request({
    data: payload,
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
    url: '/api/v2/quality-gates/evaluate',
  })
}

export const apiConstructor = (baseIntakeUrl: string, apiKey: string, appKey: string) => {
  const serviceRequest = getRequestBuilder({baseUrl: baseIntakeUrl, apiKey, appKey})

  return {
    evaluateGateRules: evaluateGateRules(serviceRequest),
  }
}
