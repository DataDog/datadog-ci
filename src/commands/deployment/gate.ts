import {isAxiosError} from 'axios'
import chalk from 'chalk'
import {Command, Option} from 'clipanion'
import * as t from 'typanion'

import {FIPS_ENV_VAR, FIPS_IGNORE_ERROR_ENV_VAR} from '../../constants'
import {toBoolean} from '../../helpers/env'
import {enableFips} from '../../helpers/fips'
import {ICONS} from '../../helpers/formatting'
import {Logger, LogLevel} from '../../helpers/logger'
import {getApiHostForSite} from '../../helpers/utils'

import {apiConstructor} from './api'
import {APIHelper, GateEvaluationRequest} from './interfaces'

type CommandResult = 'PASS' | 'FAIL'

/**
 * This command allows to evaluate a deployment gate in Datadog.
 * It handles the entire process of requesting a gate evaluation and polling for results
 * The command will exit with status 0 when the gate passes and status 1 otherwise.
 */
export class DeploymentGateCommand extends Command {
  public static paths = [['deployment', 'gate']]

  public static usage = Command.Usage({
    category: 'CI Visibility',
    description: 'Evaluate deployment gates in Datadog.',
    details: `
      This command allows to evaluate a deployment gate in Datadog.
      The command will exit with status 0 when the gate passes and status 1 otherwise.
    `,
    examples: [
      [
        'Evaluate a deployment gate for payments-backend service in prod environment',
        'datadog-ci deployment gate --service payments-backend --env prod',
      ],
      [
        'Evaluate a deployment gate with custom timeout',
        'datadog-ci deployment gate --service payments-backend --env prod --timeout 7200',
      ],
      [
        'Evaluate a deployment gate and fail if an error occurs',
        'datadog-ci deployment gate --service payments-backend --env prod --fail-on-error',
      ],
      [
        'Evaluate a deployment gate with version and APM primary tag',
        'datadog-ci deployment gate --service payments-backend --env prod --version 1.2.3 --apm-primary-tag region:us-central-1',
      ],
    ],
  })

  // Required parameters
  private service = Option.String('--service', {
    description: 'The service name (e.g. payments-backend)',
    validator: t.isString(),
  })
  private env = Option.String('--env', {
    description: 'The environment name (e.g. prod, staging)',
    validator: t.isString(),
  })

  // Optional parameters
  private identifier = Option.String('--identifier', {
    description: 'The deployment identifier (defaults to "default")',
  })
  private version = Option.String('--version', {
    description: 'The deployment version (required for gates with faulty deployment detection rules)',
  })
  private apmPrimaryTag = Option.String('--apm-primary-tag', {
    description: 'The APM primary tag (only for gates with faulty deployment detection rules)',
  })
  private timeout = Option.String('--timeout', '10800', {
    description: 'Maximum amount of seconds to wait for the script execution in seconds (default: 10800 = 3 hours)',
    validator: t.isString(),
  })
  private failOnError = Option.Boolean('--fail-on-error', false, {
    description:
      'When true, the script will consider the gate as failed when timeout is reached or unexpected errors occur calling the Datadog APIs',
  })

  // FIPS options
  private fips = Option.Boolean('--fips', false)
  private fipsIgnoreError = Option.Boolean('--fips-ignore-error', false)

  private config = {
    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
    appKey: process.env.DATADOG_APP_KEY || process.env.DD_APP_KEY,
    fips: toBoolean(process.env[FIPS_ENV_VAR]) ?? false,
    fipsIgnoreError: toBoolean(process.env[FIPS_IGNORE_ERROR_ENV_VAR]) ?? false,
  }

  private logger: Logger = new Logger((s: string) => this.context.stdout.write(s), LogLevel.INFO)
  private pollingInterval = 15000
  private startTime: number = Date.now()

  public async execute() {
    enableFips(this.fips || this.config.fips, this.fipsIgnoreError || this.config.fipsIgnoreError)

    if (!this.service) {
      this.logger.error(chalk.red(`${ICONS.FAILED} Missing required parameter: ${chalk.bold('--service')}`))

      return 1
    }

    if (!this.env) {
      this.logger.error(chalk.red(`${ICONS.FAILED} Missing required parameter: ${chalk.bold('--env')}`))

      return 1
    }

    const timeoutSeconds = parseInt(this.timeout, 10)
    if (isNaN(timeoutSeconds) || timeoutSeconds <= 0) {
      this.logger.error(
        chalk.red(`${ICONS.FAILED} Invalid ${chalk.bold('--timeout')} value. Must be a positive integer.`)
      )

      return 1
    }

    if (!this.config.apiKey) {
      this.logger.error(
        `${ICONS.FAILED} Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold(
          'DD_API_KEY'
        )} are in your environment.`
      )

      return 1
    }

    if (!this.config.appKey) {
      this.logger.error(
        `${ICONS.FAILED} Neither ${chalk.red.bold('DATADOG_APP_KEY')} nor ${chalk.red.bold(
          'DD_APP_KEY'
        )} are in your environment.`
      )

      return 1
    }

    this.logger.info('Starting deployment gate evaluation with parameters:')
    this.logger.info(`\tService: ${this.service}`)
    this.logger.info(`\tEnvironment: ${this.env}`)
    if (this.identifier) {
      this.logger.info(`\tIdentifier: ${this.identifier}`)
    }
    if (this.version) {
      this.logger.info(`\tVersion: ${this.version}`)
    }
    if (this.apmPrimaryTag) {
      this.logger.info(`\tAPM Primary Tag: ${this.apmPrimaryTag}`)
    }
    this.logger.info(`\tTimeout: ${timeoutSeconds} seconds`)
    this.logger.info(`\tFail on error: ${this.failOnError ? 'true' : 'false'}\n`)

    let result: CommandResult

    try {
      const api = this.getApiHelper(this.config.apiKey, this.config.appKey)
      const evaluationRequest = this.buildEvaluationRequest()

      const evaluationId = await this.requestGateEvaluation(api, evaluationRequest)

      result = await this.pollForEvaluationResults(api, evaluationId, timeoutSeconds)
    } catch (error) {
      this.logger.error(`Deployment gate evaluation failed: ${error instanceof Error ? error.message : String(error)}`)

      if (isAxiosError(error)) {
        if (error.response?.status && error.response.status >= 400 && error.response.status < 500) {
          this.logger.error(`${ICONS.FAILED} Request failed with client error, exiting with status 1`)

          return 1
        }
      }

      result = this.getResultForDatadogError()
    }

    return result === 'PASS' ? 0 : 1
  }

  private getApiHelper(apiKey: string, appKey: string): APIHelper {
    const site = process.env.DATADOG_SITE || process.env.DD_SITE || 'datadoghq.com'
    const baseAPIURL = `https://${getApiHostForSite(site)}`

    return apiConstructor(baseAPIURL, apiKey, appKey)
  }

  private buildEvaluationRequest(): GateEvaluationRequest {
    const request: GateEvaluationRequest = {
      service: this.service || '',
      env: this.env || '',
    }

    if (this.identifier) {
      request.identifier = this.identifier
    }

    if (this.version) {
      request.version = this.version
    }

    if (this.apmPrimaryTag) {
      request.apm_primary_tag = this.apmPrimaryTag
    }

    return request
  }

  private async requestGateEvaluation(api: APIHelper, request: GateEvaluationRequest): Promise<string> {
    this.logger.info('Requesting gate evaluation...')

    try {
      const response = await api.requestGateEvaluation(request)
      const evaluationId = response.data.data.attributes.evaluation_id
      this.logger.info(chalk.green(`Gate evaluation started successfully. Evaluation ID: ${evaluationId}\n`))

      return evaluationId
    } catch (error) {
      if (isAxiosError(error) && error.response?.status) {
        this.logger.error(`Request failed with client error: ${error.response.status} ${error.response.statusText}`)
      } else {
        this.logger.error(
          `Could not start gate evaluation with unknown error: ${
            error instanceof Error ? error.message : String(error)
          }`
        )
      }

      throw error
    }
  }

  private async pollForEvaluationResults(
    api: APIHelper,
    evaluationId: string,
    timeoutSeconds: number
  ): Promise<CommandResult> {
    this.logger.info('Waiting for gate evaluation results...')

    const maxWaitTime = timeoutSeconds * 1000
    let timePassed = Date.now() - this.startTime
    let result: CommandResult | undefined

    while (timePassed < maxWaitTime) {
      const remainingTime = maxWaitTime - timePassed
      const waitTime = Math.min(this.pollingInterval, remainingTime)
      const waitTimeInSeconds = Math.floor(waitTime / 1000)

      result = await this.getEvaluationResult(api, evaluationId)
      if (result) {
        return result
      }

      this.logger.info(`\tRetrying in ${waitTimeInSeconds}s...`)
      await new Promise((resolve) => setTimeout(resolve, waitTime))

      timePassed = Date.now() - this.startTime
    }

    // The block above may not have run the last time, so we need to check again
    result = await this.getEvaluationResult(api, evaluationId)
    if (result) {
      return result
    }

    this.logger.warn(
      `${ICONS.WARNING} Timeout reached (${timeoutSeconds} seconds). Gate evaluation did not complete in time.`
    )

    return this.getResultForDatadogError()
  }

  private async getEvaluationResult(api: APIHelper, evaluationId: string): Promise<CommandResult | undefined> {
    try {
      const response = await api.getGateEvaluationResult(evaluationId)
      const status = response.data.data.attributes.gate_status

      switch (status) {
        case 'pass':
          this.logger.info(chalk.green(`\t${ICONS.SUCCESS} Gate evaluation passed`))

          return 'PASS'
        case 'fail':
          this.logger.info(chalk.red(`\t${ICONS.FAILED} Gate evaluation failed`))

          return 'FAIL'
        case 'in_progress':
          const rules = response.data.data.attributes.rules
          const totalRules = rules.length
          const completedRules = rules.filter((rule) => rule.status !== 'in_progress').length

          this.logger.info(`\tGate evaluation in progress (${completedRules}/${totalRules} rules completed)`)
          break

        default:
          this.logger.warn(`Unknown gate evaluation status: ${status}`)
      }
    } catch (error) {
      this.logger.error(
        `Error polling for gate evaluation results: ${error instanceof Error ? error.message : String(error)}`
      )

      return this.getResultForDatadogError()
    }
  }

  private getResultForDatadogError(): CommandResult {
    if (this.failOnError) {
      this.logger.warn('Unexpected error happened, exiting with status 1 because --fail-on-error is enabled')

      return 'FAIL'
    }

    this.logger.warn('Unexpected error happened, exiting with status 0')

    return 'PASS'
  }
}
