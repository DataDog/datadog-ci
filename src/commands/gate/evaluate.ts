import chalk from 'chalk'
import {Command, Option} from 'clipanion'
import {v4 as uuidv4} from 'uuid'

import {getCISpanTags} from '../../helpers/ci'
import {getGitMetadata} from '../../helpers/git/format-git-span-data'
import {SpanTags} from '../../helpers/interfaces'
import {retryRequest} from '../../helpers/retry'
import {parseTags} from '../../helpers/tags'
import {getUserGitSpanTags} from '../../helpers/user-provided-git'

import {apiConstructor} from './api'
import {APIHelper, EvaluationResponse, Payload} from './interfaces'
import {
  renderEvaluationResponse,
  renderGateEvaluationInput,
  renderGateEvaluationError,
  renderEvaluationRetry,
  renderWaiting,
} from './renderer'
import {getBaseIntakeUrl, is4xxError, is5xxError, parseScope} from './utils'

export class GateEvaluateCommand extends Command {
  public static paths = [['gate', 'evaluate']]

  public static usage = Command.Usage({
    category: 'CI Visibility',
    description: 'Evaluate Quality Gates rules in Datadog.',
    details: `
      This command will evaluate the matching quality gate rules in Datadog.\n
      See README for details.
    `,
    examples: [
      ['Evaluate matching quality gate rules in Datadog', 'datadog-ci gate evaluate'],
      [
        'Evaluate matching quality gate rules in Datadog, failing if no rules were found',
        'datadog-ci gate evaluate --fail-on-empty',
      ],
      [
        'Evaluate matching quality gate rules in Datadog, failing if Datadog is not available',
        'datadog-ci gate evaluate --fail-if-unavailable',
      ],
      [
        'Evaluate matching quality gate rules in Datadog and add extra scope',
        'datadog-ci gate evaluate --scope team:backend',
      ],
      [
        'Evaluate matching quality gate rules in Datadog and add extra tags',
        'datadog-ci gate evaluate --tags team:frontend',
      ],
      [
        'Evaluate matching quality gate rules in Datadog from the datadoghq.eu site',
        'DATADOG_SITE=datadoghq.eu datadog-ci gate evaluate',
      ],
      ['Evaluate matching quality gate rules in Datadog without waiting', 'datadog-ci gate evaluate --no-wait'],
    ],
  })

  private dryRun = Option.Boolean('--dry-run', false)
  private failOnEmpty = Option.Boolean('--fail-on-empty', false)
  private failIfUnavailable = Option.Boolean('--fail-if-unavailable', false)
  private noWait = Option.Boolean('--no-wait', false)
  private userScope = Option.Array('--scope')
  private tags = Option.Array('--tags')

  private config = {
    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
    appKey: process.env.DATADOG_APP_KEY || process.env.DD_APP_KEY,
    envVarTags: process.env.DD_TAGS,
  }

  private waitingTime = 30000 // 30 seconds

  public async execute() {
    const api = this.getApiHelper()
    const spanTags = await this.getSpanTags()
    const userScope = this.userScope ? parseScope(this.userScope) : {}

    const payload = {
      requestId: uuidv4(),
      spanTags,
      userScope,
      options: {
        dryRun: this.dryRun,
      },
    }

    return this.evaluateRules(api, payload)
  }

  private getApiHelper(): APIHelper {
    if (!this.config.apiKey) {
      this.context.stdout.write(
        `Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold('DD_API_KEY')} is in your environment.\n`
      )
      throw new Error('API key is missing')
    }

    if (!this.config.appKey) {
      this.context.stdout.write(`Missing ${chalk.red.bold('DATADOG_APP_KEY')} in your environment.\n`)
      throw new Error('APP key is missing')
    }

    return apiConstructor(getBaseIntakeUrl(), this.config.apiKey, this.config.appKey)
  }

  private async getSpanTags(): Promise<SpanTags> {
    const ciSpanTags = getCISpanTags()
    const gitSpanTags = await getGitMetadata()
    const userGitSpanTags = getUserGitSpanTags()

    const envVarTags = this.config.envVarTags ? parseTags(this.config.envVarTags.split(',')) : {}
    const cliTags = this.tags ? parseTags(this.tags) : {}

    return {
      ...gitSpanTags,
      ...ciSpanTags,
      ...userGitSpanTags,
      ...cliTags,
      ...envVarTags,
    }
  }

  private async evaluateRules(api: APIHelper, evaluateRequest: Payload): Promise<number> {
    if (this.shouldWait()) {
      this.context.stdout.write(renderWaiting())
      await this.delay(this.waitingTime)
    }

    this.context.stdout.write(renderGateEvaluationInput(evaluateRequest))

    return retryRequest(
      () => api.evaluateGateRules(evaluateRequest, this.context.stdout.write.bind(this.context.stdout)),
      {
        onRetry: (e, attempt) => {
          this.context.stderr.write(renderEvaluationRetry(attempt, e))
        },
        retries: 5,
      }
    )
      .then((response) => {
        return this.handleEvaluationSuccess(response.data.data.attributes)
      })
      .catch((error) => {
        return this.handleEvaluationError(error)
      })
  }

  private handleEvaluationSuccess(evaluationResponse: EvaluationResponse): number {
    this.context.stdout.write(renderEvaluationResponse(evaluationResponse))

    if (evaluationResponse.status === 'failed' || (evaluationResponse.status === 'empty' && this.failOnEmpty)) {
      return 1
    }

    return 0
  }

  private handleEvaluationError(error: any): number {
    this.context.stderr.write(renderGateEvaluationError(error, this.failIfUnavailable))
    if (is4xxError(error) || (is5xxError(error) && this.failIfUnavailable)) {
      return 1
    }

    return 0
  }

  private async delay(ms: number) {
    await new Promise((resolve) => setTimeout(resolve, ms))
  }

  private shouldWait(): boolean {
    return !this.dryRun && !this.noWait
  }
}
