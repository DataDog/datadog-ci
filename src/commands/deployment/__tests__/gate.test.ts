import {createCommand, makeRunCLI} from '../../../helpers/__tests__/testing-tools'

import {DeploymentGateCommand} from '../gate'

describe('gate', () => {
  describe('execute', () => {
    const runCLI = makeRunCLI(DeploymentGateCommand, ['deployment', 'gate'], {skipResetEnv: true})

    let originalEnv: NodeJS.ProcessEnv
    beforeEach(() => {
      originalEnv = {...process.env}
    })

    afterEach(() => {
      process.env = originalEnv
    })

    test('should fail if service is not provided', async () => {
      const {context, code} = await runCLI(['--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Missing required parameter: --service')
    })

    test('should fail if env is not provided', async () => {
      const {context, code} = await runCLI(['--service', 'test-service'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Missing required parameter: --env')
    })

    test('should fail if API key is not provided', async () => {
      delete process.env.DATADOG_API_KEY
      delete process.env.DD_API_KEY

      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Neither DATADOG_API_KEY nor DD_API_KEY is in your environment')
    })

    test('should fail if APP key is not provided', async () => {
      delete process.env.DATADOG_APP_KEY
      delete process.env.DD_APP_KEY

      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Neither DATADOG_APP_KEY nor DD_APP_KEY is in your environment')
    })

    test('should fail if timeout is invalid', async () => {
      const {context, code} = await runCLI(['--service', 'test-service', '--env', 'prod', '--timeout', 'invalid'])
      expect(code).toBe(1)
      expect(context.stdout.toString()).toContain('Invalid timeout value. Must be a positive integer.')
    })
  })

  describe('buildEvaluationRequest', () => {
    test('should build basic request with required parameters', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
      })
    })

    test('should include version when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['version'] = '1.2.3'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        version: '1.2.3',
      })
    })

    test('should include apm_primary_tag when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['apmPrimaryTag'] = 'team:backend'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        apm_primary_tag: 'team:backend',
      })
    })

    test('should include both version and apm_primary_tag when provided', () => {
      const command = createCommand(DeploymentGateCommand)
      command['service'] = 'test-service'
      command['env'] = 'prod'
      command['identifier'] = 'default'
      command['version'] = '1.2.3'
      command['apmPrimaryTag'] = 'team:backend'

      const request = command['buildEvaluationRequest']()
      expect(request).toEqual({
        service: 'test-service',
        env: 'prod',
        identifier: 'default',
        version: '1.2.3',
        apm_primary_tag: 'team:backend',
      })
    })
  })
})
