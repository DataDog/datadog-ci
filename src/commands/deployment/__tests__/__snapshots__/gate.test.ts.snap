// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gate execute evaluation errors on gate evaluation request should fail when gate evaluation request fails with 500 and fail-on-error is true 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: true

Requesting gate evaluation...
Request failed with client error: 500 Internal Server Error
Deployment gate evaluation failed: Request failed with status code 500
Unexpected error happened, exiting with status 1 because --fail-on-error is enabled
"
`;

exports[`gate execute evaluation errors on gate evaluation request should pass when gate evaluation request fails with 500 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Request failed with client error: 500 Internal Server Error
Deployment gate evaluation failed: Request failed with status code 500
Unexpected error happened, exiting with status 0
"
`;

exports[`gate execute evaluation errors on gate evaluation result pass with a 500 error 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Error polling for gate evaluation results: API Error
Unexpected error happened, exiting with status 0
"
`;

exports[`gate execute evaluation errors on gate evaluation result should fail with 500 error when fail-on-error is true 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: true

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Error polling for gate evaluation results: Request failed with status code 500
Unexpected error happened, exiting with status 1 because --fail-on-error is enabled
"
`;

exports[`gate execute evaluation errors on gate evaluation result should not fail when gate evaluation result is invalid 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
Unknown gate evaluation status: unexpected
	Retrying in 15s...
	✅ Gate evaluation passed
"
`;

exports[`gate execute successful evaluation should fail when gate evaluation fails 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	❌ Gate evaluation failed
"
`;

exports[`gate execute successful evaluation should pass when gate evaluation passes after multiple in_progress calls 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	Gate evaluation in progress (0/2 rules completed)
	Retrying in 15s...
	Gate evaluation in progress (1/2 rules completed)
	Retrying in 15s...
	✅ Gate evaluation passed
"
`;

exports[`gate execute successful evaluation should succeed when gate evaluation passes on first poll 1`] = `
"Starting deployment gate evaluation with parameters:
	Service: test-service
	Environment: prod
	Timeout: 10800 seconds
	Fail on error: false

Requesting gate evaluation...
Gate evaluation started successfully. Evaluation ID: test-evaluation-id

Waiting for gate evaluation results...
	✅ Gate evaluation passed
"
`;
